### DISCUSSION TITLE
Hello World Example for Phoenix 1.7

### DISCUSSION THREAD

**Discussion by wintermeyer** [maennchen/iona#30] [Q&A]
Does anybody have a step by step how-to for a Hello World example with Phoenix 1.7? Starting with a fresh Phoenix app. That would help me and others a lot. Let's assume I do a `mix phx.gen.html Accounts User users name:string age:integer` and want to generate not just a HTML `:show` of that user but also download a PDF when the `.pdf` file is requested. Thank you! 🙏

BTW: Great project! Solves a pain point for my https://feriendaten.de open-source project and probably others.

**Comment by maennchen** [comment]
Hi @wintermeyer 

That should not be very complicated:

## Dedicated Action in controller

* Create a `.tex.eex` file in `priv` containing `<%= @user.name %>` etc.
* add an action to your controller

`lib/controller/user_controller.ex`

```elixir
@spec show_pdf(conn :: Plug.Conn.t(), params :: map()) :: Plug.Conn.t()
def show_pdf(conn, %{"user_id" => user_id}) do
  %User{name: name} = user = UserContext.get_user!(user_id)

  pdf = %{user: user}
  |> Iona.template!(path: Application.app_dir(:acme, "/priv/path/to/template.tex.eex"))
  |> Iona.to(:pdf)

  conn
  |> put_resp_header("content-type", "application/pdf")
  |> put_resp_header("content-disposition", "attachment;filename=#{name}.pdf")
  |> send_resp(:ok, pdf)
end
```

## Content Type Switch

In theory, you could also allow PDF access (same as JSON / HTML) to your controller and move your pdf generation out of the controller.

in the router (either add to the existing pipeline or add a new one)

`lib/router.ex`

```elixir
plug :accepts, ["html", "pdf"]
```

`lib/controller/user_controller.ex`

```elixir
use Phoenix.Controller,
  namespace: Acme,
  formats: [:html, :pdf],
  layouts: [html: Acme.Layouts]

# ...

@spec show(conn :: Plug.Conn.t(), params :: map()) :: Plug.Conn.t()
def show(conn, %{"user_id" => user_id}) do
  user = UserContext.get_user!(user_id)

  render(conn, :show, user: user)
end
```

`lib/controller/user_pdf.ex`

```elixir
defmodule Acme.UserPDF do
  def show(assigns) do
    pdf = assigns
    |> Iona.template!(path: Application.app_dir(:acme, "/priv/path/to/template.tex.eex"))
    |> Iona.to(:pdf)
  end
end

```

(All code in this answer is entirely untested. It probably contains bugs.)
